src/ripple/rpc/handlers/WalletPropose.cpp

```diff
@@ -85,48 +85,69 @@ Json::Value walletPropose (Json::Value const& params)
         keyType = keyTypeFromString (
             params[jss::key_type].asString());
 
-        if (keyType == KeyType::invalid)
+        if (!keyType)
             return rpcError(rpcINVALID_PARAMS);
     }
 
-    if (params.isMember (jss::passphrase) ||
-        params.isMember (jss::seed) ||
-        params.isMember (jss::seed_hex))
+    // ripple-lib encodes seed used to generate an Ed25519 wallet in a
+    // non-standard way. While we never encode seeds that way, we try
+    // to detect such keys to avoid user confusion.
+    if(seed = RPC::parseRippleLibSeed(params))
     {
-        Json::Value err;
-        seed = RPC::getSeedFromRPC (params, err);
-        if (!seed)
-            return err;
+        rippleLibSeed = true;
+
+        // If the user *explicitly* requests a key type other than
+        // Ed25519 we return an error.
+        if (keyType.value_or(KeyType::ed25519) != KeyType::ed25519)
+            return rpcError(rpcBAD_SEED);
+
+        keyType = KeyType::ed25519;
     }
-    else
+
+    if (!seed)
     {
-        seed = randomSeed ();
+        if (params.isMember(jss::passphrase) ||
+            params.isMember(jss::seed) ||
+            params.isMember(jss::seed_hex))
+        {
+            Json::Value err;
+
+            seed = RPC::getSeedFromRPC(params, err);
+
+            if (!seed)
+                return err;
+        }
+        else
+        {
+            seed = randomSeed();
+        }
     }
 
-    auto const publicKey = generateKeyPair (keyType, *seed).first;
+    if (!keyType)
+        keyType = KeyType::secp256k1;
+
+    auto const publicKey = generateKeyPair (*keyType, *seed).first;
 
     Json::Value obj (Json::objectValue);
 
-    auto const seed1751 = seedAs1751 (*seed);
     auto const seedHex = strHex (*seed);
     auto const seedBase58 = toBase58 (*seed);
 
     obj[jss::master_seed] = seedBase58;
```

Comment:
**IMPORTANT THING TO POINT OUT:**

While we parse the `ripple-lib` encoded Ed25519 key correctly, we return the "canonical" seed encoding. Should that be changed? And, more generally, should we just accept that that an encoded seed represents the key type and the seed used to generate the wallet?

Reply:
It seems we should return an error upon detecting a type mismatch especially since we already reject upon an invalid `keyType`.

Reply:
We do.